tests/sync/test_client.py

```diff
@@ -115,3 +114,96 @@ def test_should_discover_ip():
     hostport = client.hostport
 
     assert '0.0.0.0:0' != hostport
+
+
+@pytest.mark.gen_test
+def test_sync_register_json():
+    sync_client = TChannel('test-client')
+
+    @sync_client.json.register
+    def hello(request):
+        return {}
+
+    sync_client.listen()
+
+    from tchannel import TChannel as AsyncTchannel
+    async_client = AsyncTchannel('s')
+    with pytest.raises(BadRequestError):
+        yield async_client.json(
+            service='test-client',
+            endpoint='hello',
+            hostport=sync_client.hostport
+        )
+
+    sync_client.json.register(hello)
+    with pytest.raises(BadRequestError):
```

Comment:
Can you not duplicate these tests like this? It's not very difficult to use fixtures for this. Declare these fixtures:

- scheme: Just a string value. One of json, thrift, raw.
- register_endpoints(scheme): returns a function that takes a tchannel and registers endpoints against it. could optionally be parameterized over whether you use a decorator or a manual `.register` call.
- call(scheme): Returns a function that makes a request to endpoint defined in register_endpoints for the specific scheme.

Then write two separate tests, both using register_endpoints and call: one with a sync tchannel, one with an async tchannel. This allows you to test the 3 * 2 * 2 approaches with two tests rather than 12.